File: 2025-11/2025-11-19/markdownHeadingConverter.js
function convert(heading) {
  const matchResult = /^ *(?<headingLevel>#{1,6}) +(?<headingText>.+)/.exec(heading)?.groups;
  
  if(!matchResult)
    return "Invalid format";

  const { headingLevel, headingText } = matchResult;

  return `<h${headingLevel.length}>${headingText}</h${headingLevel.length}>`;
}

console.log(convert("# My level 1 heading"));
console.log(convert("My heading"));
console.log(convert("##### My level 5 heading"));
console.log(convert("#My heading"));
console.log(convert("  ###  My level 3 heading"));
console.log(convert("####### My level 7 heading"));
console.log(convert("## My #2 heading"));

// --- TEST SUITE ---

const testsText = `
1. convert("# My level 1 heading") should return "<h1>My level 1 heading</h1>".
2. convert("My heading") should return "Invalid format".
3. convert("##### My level 5 heading") should return "<h5>My level 5 heading</h5>".
4. convert("#My heading") should return "Invalid format".
5. convert("  ###  My level 3 heading") should return "<h3>My level 3 heading</h3>".
6. convert("####### My level 7 heading") should return "Invalid format".
7. convert("## My #2 heading") should return "<h2>My #2 heading</h2>".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+)\.$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("---------------------------");
  console.log("🧪 Starting Verification...");
  console.log("---------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    if (functionCallOutput === testOutput) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("-----------------------------");

  if (allPassed)
    console.log("🎉 SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️ WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

runTests(testData);

File: 2025-11/2025-11-19/markdown_heading_converter.py
import re

def convert(heading):
    match_result = re.match(" *(?P<heading_level>#{1,6}) +(?P<heading_text>.+)", heading)
    
    if(not match_result):
        return "Invalid format"

    heading_level, heading_text = match_result.groups()

    return f"<h{len(heading_level)}>{heading_text}</h{len(heading_level)}>"

print(convert("# My level 1 heading"))
print(convert("My heading"))
print(convert("##### My level 5 heading"))
print(convert("#My heading"))
print(convert("  ###  My level 3 heading"))
print(convert("####### My level 7 heading"))
print(convert("## My #2 heading"))

tests_text = '''
1. convert("# My level 1 heading") should return "<h1>My level 1 heading</h1>".
2. convert("My heading") should return "Invalid format".
3. convert("##### My level 5 heading") should return "<h5>My level 5 heading</h5>".
4. convert("#My heading") should return "Invalid format".
5. convert("  ###  My level 3 heading") should return "<h3>My level 3 heading</h3>".
6. convert("####### My level 7 heading") should return "Invalid format".
7. convert("## My #2 heading") should return "<h2>My #2 heading</h2>".
'''

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+)\.$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("---------------------------")
    print("🧪 Starting Verification...")
    print("---------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)